packages/vitest/src/node/pool.ts

```diff
@@ -39,51 +42,69 @@ export function createPool(ctx: Vitest): WorkerPool {
     execArg.startsWith('--cpu-prof') || execArg.startsWith('--heap-prof'),
   )
 
-  const options: TinypoolOptions = {
-    filename: workerPath,
-    // TODO: investigate further
-    // It seems atomics introduced V8 Fatal Error https://github.com/vitest-dev/vitest/issues/1191
-    useAtomics: ctx.config.useAtomics ?? false,
-
-    maxThreads,
-    minThreads,
-
+  const options: ProcessOptions = {
     execArgv: ctx.config.deps.registerNodeLoader
       ? [
           ...execArgv,
           '--require',
           suppressLoaderWarningsPath,
           '--experimental-loader',
           loaderPath,
-          ...conditions,
+          ...execArgv,
         ]
       : [
           ...execArgv,
           ...conditions,
         ],
+    env: {
+      TEST: 'true',
+      VITEST: 'true',
+      NODE_ENV: ctx.config.mode || 'test',
+      VITEST_MODE: ctx.config.watch ? 'WATCH' : 'RUN',
+      ...process.env,
+      ...ctx.config.env,
+    },
+  }
+
+  if (!ctx.config.threads)
+    return createChildProcessPool(ctx, options)
+  return createThreadsPool(ctx, options)
+}
+
+export function createThreadsPool(ctx: Vitest, { execArgv, env }: ProcessOptions): WorkerPool {
+  const threadsCount = ctx.config.watch
+    ? Math.max(Math.floor(cpus().length / 2), 1)
+    : Math.max(cpus().length - 1, 1)
+
+  const maxThreads = ctx.config.maxThreads ?? threadsCount
+  const minThreads = ctx.config.minThreads ?? threadsCount
+
+  const options: TinypoolOptions = {
+    filename: workerPath,
+    // TODO: investigate further
+    // It seems atomics introduced V8 Fatal Error https://github.com/vitest-dev/vitest/issues/1191
+    useAtomics: ctx.config.useAtomics ?? false,
+
+    maxThreads,
+    minThreads,
+
+    execArgv,
   }
 
   if (ctx.config.isolate) {
     options.isolateWorkers = true
     options.concurrentTasksPerWorker = 1
   }
 
-  if (!ctx.config.threads) {
+  if (ctx.config.singleThread) {
     options.concurrentTasksPerWorker = 1
     options.maxThreads = 1
     options.minThreads = 1
   }
 
   ctx.coverageProvider?.onBeforeFilesRun?.()
 
-  options.env = {
-    TEST: 'true',
-    VITEST: 'true',
-    NODE_ENV: ctx.config.mode || 'test',
-    VITEST_MODE: ctx.config.watch ? 'WATCH' : 'RUN',
-    ...process.env,
-    ...ctx.config.env,
-  }
+  options.env = env
```

Comment:
The C8 provider's `onBeforeFilesRun` modifies main thread's `process.env` which is then copied passed to workers. Use following to get coverage working again.

```ts
  ctx.coverageProvider?.onBeforeFilesRun?.()

  options.env = {
    ...env,
    ...process.env
  }
```

This is quite hacky as there is such side effect, and I've been planning to make it more visible. You are currently seeing effects of this bug/feature: https://github.com/nodejs/node/issues/46378

Reply:
Hm, I see... We should probably rename the hook then 🤔"before field run" is not really what is happening here I think 🤔